src/provider/provider_file_memory.c

```diff
@@ -0,0 +1,620 @@
+    }
+
+    file_memory_provider_t *file_provider =
+        umf_ba_global_alloc(sizeof(file_memory_provider_t));
+    if (!file_provider) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    memset(file_provider, 0, sizeof(*file_provider));
+
+    ret = file_translate_params(in_params, file_provider);
+    if (ret != UMF_RESULT_SUCCESS) {
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err_free_file_provider;
+    }
+
+    if (util_copy_path(in_params->path, file_provider->path, PATH_MAX)) {
+        goto err_free_file_provider;
+    }
+
+    file_provider->fd = os_file_open_or_create(in_params->path, FILE_INIT_SIZE);
+    if (file_provider->fd == -1) {
+        LOG_ERR("cannot open the file: %s", in_params->path);
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err_free_file_provider;
+    }
+
+    if (os_get_file_size(file_provider->fd, &file_provider->size_fd)) {
+        LOG_ERR("cannot get size of the file: %s", in_params->path);
+        ret = UMF_RESULT_ERROR_UNKNOWN;
+        goto err_close_fd;
+    }
+
+    LOG_DEBUG("size of the file %s is: %zu", in_params->path,
+              file_provider->size_fd);
+
+    if (util_mutex_init(&file_provider->lock_fd) == NULL) {
+        LOG_ERR("lock init failed");
+        ret = UMF_RESULT_ERROR_UNKNOWN;
+        goto err_close_fd;
+    }
+
+    file_provider->fd_offset_map = critnib_new();
+    if (!file_provider->fd_offset_map) {
+        LOG_ERR("creating file descriptor offset map failed");
+        ret = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_mutex_destroy_not_free;
+    }
+
+    *provider = file_provider;
+
+    return UMF_RESULT_SUCCESS;
+
+err_mutex_destroy_not_free:
+    util_mutex_destroy_not_free(&file_provider->lock_fd);
+err_close_fd:
+    utils_close_fd(file_provider->fd);
+err_free_file_provider:
+    umf_ba_global_free(file_provider);
+    return ret;
+}
+
+static void file_finalize(void *provider) {
+    if (provider == NULL) {
+        assert(0);
+        return;
+    }
+
+    file_memory_provider_t *file_provider = provider;
+    util_mutex_destroy_not_free(&file_provider->lock_fd);
+    critnib_delete(file_provider->fd_offset_map);
+    umf_ba_global_free(file_provider);
+}
+
+static umf_result_t file_get_min_page_size(void *provider, void *ptr,
+                                           size_t *page_size);
+
+static inline void assert_is_page_aligned(uintptr_t ptr, size_t page_size) {
+    assert((ptr & (page_size - 1)) == 0);
+    (void)ptr;       // unused in Release build
+    (void)page_size; // unused in Release build
+}
+
+static int file_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
+                             size_t page_size, int prot, int flag, int fd,
+                             size_t *size_fd, os_mutex_t *lock_fd,
+                             void **out_addr, size_t *offset_fd,
+                             size_t *old_offset_fd) {
```

Comment:
12 arguments to single function? Can we reduce it?
Also this function is almost ideal copy of this in os memory provider.  (and in #643 it was promissed to be refactored in separate PR)

Refactor ideas:
1. Abstract it to common, as it is needed in multiple places. 
2. Split it to multiple separate functions (i.e. One to grow file, second for mmap)

Reply:
Done

Reply:
Done.

@lplewa ping